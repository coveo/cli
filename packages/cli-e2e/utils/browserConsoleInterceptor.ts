import {CDPSession, Page} from 'puppeteer';
import type Protocol from 'devtools-protocol';
import {FileLogger} from './filelogger';

type ConsoleAPICalledEventType = Protocol.Runtime.ConsoleAPICalledEvent['type'];
type ConsoleAPICalledEvent = Protocol.Runtime.ConsoleAPICalledEvent;

export class BrowserConsoleInterceptor {
  private _client: CDPSession | null = null;
  private fileLogger: FileLogger;
  public interceptedMessages: string[] = [];

  constructor(private page: Page, private interceptorName: string) {
    this.fileLogger = new FileLogger(
      `console${
        this.interceptorName ? '-' + this.interceptorName : ''
      }`.replace(/[^\w\d]/g, '-')
    );
  }

  private async init() {
    this._client = await this.page.target().createCDPSession();
    await this._client.send('Runtime.enable');
    return this._client;
  }

  private async getClient(): Promise<CDPSession> {
    return this._client || (await this.init());
  }

  public async startSession(
    typesToIntercept: ConsoleAPICalledEventType[] = ['error', 'warning']
  ) {
    const client = await this.getClient();
    client.on('Runtime.consoleAPICalled', (message: ConsoleAPICalledEvent) =>
      this.onClient(message, typesToIntercept)
    );
  }

  private onClient(
    message: ConsoleAPICalledEvent,
    typesToIntercept: ConsoleAPICalledEventType[]
  ) {
    if (typesToIntercept.some((type) => type === message.type)) {
      message.args.forEach((arg) => {
        if (arg.description) {
          this.interceptedMessages.push(arg.description);
          this.logMessage(arg.description, message.type);
        }
      });
    }
  }

  private logMessage(message: string, type: string) {
    this.fileLogger[type === 'error' ? 'stderr' : 'stdout'].write(message);
  }

  public async endSession() {
    const client = await this.getClient();
    client.removeAllListeners('Runtime.consoleAPICalled');
    this.interceptedMessages = [];
  }
}
