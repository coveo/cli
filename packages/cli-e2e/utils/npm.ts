export const npmPathEnvVar = 'E2E_NPM_PATH';
export const npmCachePathEnvVar = 'E2E_NPM_CACHE_PATH';
export const npm = () => ['node', process.env[npmPathEnvVar]!];
