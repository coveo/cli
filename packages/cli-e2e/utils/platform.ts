import {HTTPRequest} from 'puppeteer';

export function isSearchRequest(request: HTTPRequest) {
  return request.url().startsWith(
    // TODO: CDX-98: URL should vary in fonction of the targeted environment.
    'https://platformdev.cloud.coveo.com/rest/search/v2?organizationId'
  );
}
