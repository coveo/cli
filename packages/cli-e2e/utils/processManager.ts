import {
  ChildProcessWithoutNullStreams,
  SpawnOptionsWithoutStdio,
  spawn as nativeSpawn,
} from 'child_process';
import fkill from 'fkill';
export class ProcessManager {
  private processes: Set<ChildProcessWithoutNullStreams>;
  public constructor() {
    this.processes = new Set<ChildProcessWithoutNullStreams>();
  }

  public spawn(
    command: string,
    args?: ReadonlyArray<string>,
    options?: SpawnOptionsWithoutStdio
  ): ChildProcessWithoutNullStreams {
    const process = nativeSpawn(command, args, {
      detached: true,
      ...options,
    });
    process.on('exit', this.onExit(process));
    this.processes.add(process);
    return process;
  }

  private onExit = (process: ChildProcessWithoutNullStreams) => () => {
    this.processes.delete(process);
  };

  public async killAllProcesses() {
    const pids: Array<number> = Array.from(this.processes.values())
      .map(
        process.platform === 'win32'
          ? (process) => process.pid || 0
          : (process) => -(process.pid || 0)
      )
      .filter((pid) => pid !== 0);
    await fkill(pids, {tree: true, force: true, silent: true});
  }
}
