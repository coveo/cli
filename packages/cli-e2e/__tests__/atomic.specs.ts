import type {HTTPRequest, Browser, Page} from 'puppeteer';
import {captureScreenshots, getNewBrowser, openNewPage} from '../utils/browser';
import {answerPrompt, getProjectPath, setupUIProject} from '../utils/cli';
import {isSearchRequestOrResponse} from '../utils/platform';
import {ProcessManager} from '../utils/processManager';
import {Terminal} from '../utils/terminal/terminal';
import {BrowserConsoleInterceptor} from '../utils/browserConsoleInterceptor';
import {npm} from '../utils/npm';
import {jwtTokenPattern} from '../utils/matcher';
import {EOL} from 'os';
import {DummyServer} from '../utils/server';
import {loginWithApiKey} from '../utils/login';
import {join, resolve} from 'path';
import {hashElement} from 'folder-hash';
import {renameSync, rmSync} from 'fs';
import retry from 'async-retry';

interface BuildAppOptions {
  id: string;
  pageId?: string;
  pageName?: string;
  skipInstall?: boolean;
}

describe('ui:create:atomic', () => {
  const searchPageEndpoint = 'http://localhost:8888';
  const tokenServerEndpoint = 'http://localhost:8888/.netlify/functions/token';
  const searchInterfaceSelector = 'atomic-search-interface';
  let normalizedProjectDir = '';

  const normalizeProjectDirectory = async (
    buildAppOptions: BuildAppOptions
  ) => {
    const originalProjectDir = resolve(
      join(getProjectPath(getProjectName(buildAppOptions.id)))
    );
    normalizedProjectDir = join(originalProjectDir, '..', 'normalizedDir');
    rmSync(normalizedProjectDir, {recursive: true, force: true});
    await retry(() => {
      renameSync(originalProjectDir, normalizedProjectDir);
    });
  };

  const waitForAppRunning = (appTerminal: Terminal) =>
    appTerminal
      .when(/build finished/)
      .on('stdout')
      .do()
      .once();

  const getProjectName = (id: string) =>
    `${process.env.TEST_RUN_ID}-atomic-project-${id}`;

  const getFlagsFromOptions = (options: BuildAppOptions) =>
    options.pageId ? ['--pageId', options.pageId] : undefined;

  const selectDefaultWithCli = (buildTerminal: Terminal) => {
    buildTerminal
      .when(/Use an existing hosted search page/)
      .on('stdout')
      .do(answerPrompt(EOL))
      .once();
  };

  const selectPageWithCli = (
    options: BuildAppOptions,
    buildTerminal: Terminal
  ) => {
    const osSpecificSelector = process.platform === 'win32' ? '>' : '\u276f'; //❯
    const anotherPageSelectedMatcher = new RegExp(
      `${osSpecificSelector} (?!${options.pageName})`,
      'gm'
    );
    const expectedPageSelectedMatcher = new RegExp(
      `${osSpecificSelector} ${options.pageName}`,
      'gm'
    );
    buildTerminal
      .when(anotherPageSelectedMatcher)
      .on('stdout')
      .do(answerPrompt('\x1B[B'))
      .until(expectedPageSelectedMatcher);
    buildTerminal
      .when(expectedPageSelectedMatcher)
      .on('stdout')
      .do(answerPrompt(EOL))
      .once();
  };

  const buildApplication = async (
    processManager: ProcessManager,
    options: BuildAppOptions
  ) => {
    let output = '';

    const buildTerminal = await setupUIProject(
      processManager,
      'ui:create:atomic',
      getProjectName(options.id),
      {
        flags: getFlagsFromOptions(options),
      }
    );

    buildTerminal.orchestrator.process.stderr.on('data', (chunk: string) => {
      output += chunk;
    });
    buildTerminal.orchestrator.process.stdout.on('data', (chunk: string) => {
      output += chunk;
    });

    if (options.pageName) {
      selectPageWithCli(options, buildTerminal);
    } else if (!options.pageId) {
      selectDefaultWithCli(buildTerminal);
    }

    const streams = ['stdout', 'stderr'] as const;
    const buildTerminalExitPromise = Promise.race([
      buildTerminal.when('exit').on('process').do().once(),
      ...streams.map((stream) =>
        buildTerminal
          .when(options.skipInstall ? /Installing packages/ : /Happy hacking!/)
          .on(stream)
          .do()
          .once()
      ),
    ]);

    await Promise.allSettled(
      streams.map((stream) =>
        buildTerminal
          .when(/\(y\)/)
          .on(stream)
          .do(answerPrompt(`y${EOL}`))
          .until(buildTerminalExitPromise)
      )
    );

    return {output};
  };

  const startApplication = (
    processManager: ProcessManager,
    options: BuildAppOptions,
    debugName = 'atomic-server'
  ) => {
    const args = [...npm(), 'run', 'start'];

    return new Terminal(
      args.shift()!,
      args,
      {
        cwd: normalizedProjectDir,
      },
      processManager,
      `${debugName}-${options.id}`
    );
  };

  describe.each([
    {
      describeName: 'when using an existing pageId (pageId not specified)',
      buildAppOptions: {
        id: 'with-page-id',
        pageId: 'fffaafcc-6863-46cb-aca3-97522fcc0f5d',
        skipInstall: false,
      },
      skipBrowser: false,
    },
    {
      describeName: 'when using the default page config (pageId not specified)',
      buildAppOptions: {id: 'without-page-id', skipInstall: true},
      skipBrowser: true,
    },
    {
      describeName: 'when using an existing pageId (--pageId flag specified)',
      buildAppOptions: {
        id: 'with-page-id',
        pageId: 'fffaafcc-6863-46cb-aca3-97522fcc0f5d',
        skipInstall: true,
      },
      skipBrowser: true,
    },
    {
      describeName:
        'when using an existing pageId (using the list prompt of available pages)',
      buildAppOptions: {
        id: 'with-page-id-prompt',
        pageName: 'cli-tests-do-not-delete',
        skipInstall: true,
      },
      skipBrowser: true,
    },
  ])(
    '$describeName',
    ({
      buildAppOptions,
      skipBrowser,
    }: {
      buildAppOptions: BuildAppOptions;
      skipBrowser: boolean;
    }) => {
      const oldEnv = process.env;
      const processManagers: ProcessManager[] = [];
      let stderr: string;
      let browser: Browser;
      let page: Page;

      beforeAll(async () => {
        await loginWithApiKey(
          process.env.PLATFORM_API_KEY!,
          process.env.ORG_ID!,
          process.env.PLATFORM_ENV!
        );
        const processManager = new ProcessManager();
        processManagers.push(processManager);
        if (!skipBrowser) {
          browser = await getNewBrowser();
        }
        stderr = (await buildApplication(processManager, buildAppOptions))
          .output;
        await processManager.killAllProcesses();
        await normalizeProjectDirectory(buildAppOptions);
      }, 15 * 60e3);

      beforeEach(async () => {
        jest.resetModules();
        process.env = {...oldEnv};
        if (!skipBrowser) {
          page = await openNewPage(browser, page);
        }
      });

      afterEach(async () => {
        if (!skipBrowser) {
          await captureScreenshots(browser);
        }
      });

      afterAll(async () => {
        process.env = oldEnv;
        if (!skipBrowser) {
          await browser.close();
        }
        await Promise.all(
          processManagers.map((manager) => manager.killAllProcesses())
        );
      }, 5 * 30e3);

      it('should use the right configuration', () => {
        const message =
          buildAppOptions.pageId || buildAppOptions.pageName
            ? 'Hosted search page named'
            : 'Using the default search page template.';
        expect(stderr).toContain(message);
      });

      it('the project has been generated properly', async () => {
        expect(
          await hashElement(normalizedProjectDir, {
            folders: {
              exclude: ['**node_modules', 'dist', 'www', '.netlify'],
              ignoreRootName: true,
              ignoreBasename: true,
            },
            files: {
              include: ['*'],
              exclude: [
                '**.env',
                '**package-lock.json',
                '**CHANGELOG.md',
                '**package.json',
                'stencil.config.ts',
                '**index.html',
              ],
              ignoreRootName: true,
              ignoreBasename: true,
            },
          })
        ).toMatchSnapshot();
      });

      if (!skipBrowser) {
        describe('when the project is configured correctly', () => {
          let serverProcessManager: ProcessManager;
          let interceptedRequests: HTTPRequest[] = [];
          let consoleInterceptor: BrowserConsoleInterceptor;

          beforeAll(async () => {
            serverProcessManager = new ProcessManager();
            processManagers.push(serverProcessManager);
            const appTerminal = startApplication(
              serverProcessManager,
              buildAppOptions,
              'atomic-server-valid'
            );
            await waitForAppRunning(appTerminal);
          }, 5 * 60e3);

          beforeEach(async () => {
            consoleInterceptor = new BrowserConsoleInterceptor(
              page,
              getProjectName(buildAppOptions.id)
            );
            await consoleInterceptor.startSession();

            page.on('request', (request: HTTPRequest) => {
              interceptedRequests.push(request);
            });
          });

          afterEach(async () => {
            page.removeAllListeners('request');
            interceptedRequests = [];
            await consoleInterceptor.endSession();
          });

          afterAll(async () => {
            await serverProcessManager.killAllProcesses();
          }, 5 * 30e3);

          it.skip('should not contain console errors nor warnings', async () => {
            await page.goto(searchPageEndpoint, {
              waitUntil: 'networkidle2',
            });

            expect(consoleInterceptor.interceptedMessages).toEqual([]);
          }, 60e3);

          it('should contain a search page section', async () => {
            await page.goto(searchPageEndpoint, {
              waitUntil: 'networkidle2',
            });

            expect(await page.$(searchInterfaceSelector)).not.toBeNull();
          }, 60e3);

          it('should retrieve the search token on the page load', async () => {
            const tokenResponseListener =
              page.waitForResponse(tokenServerEndpoint);
            page.goto(searchPageEndpoint);
            await page.waitForSelector(searchInterfaceSelector);

            expect(
              JSON.parse(await (await tokenResponseListener).text())
            ).toMatchObject({
              token: expect.stringMatching(jwtTokenPattern),
            });
          }, 60e3);

          it('should send a search query when the page is loaded', async () => {
            await page.goto(searchPageEndpoint, {waitUntil: 'networkidle2'});
            await page.waitForSelector(searchInterfaceSelector);

            expect(
              interceptedRequests.some(isSearchRequestOrResponse)
            ).toBeTruthy();
          }, 60e3);
        });

        describe('when the default Stencil port is busy', () => {
          let dummyServer: DummyServer;
          let serverProcessManager: ProcessManager;

          beforeAll(async () => {
            serverProcessManager = new ProcessManager();
            processManagers.push(serverProcessManager);

            dummyServer = new DummyServer(3333);
            await dummyServer.start();

            const appTerminal = startApplication(
              serverProcessManager,
              buildAppOptions,
              'stencil-port-test'
            );
            await waitForAppRunning(appTerminal);
          }, 2 * 60e3);

          afterAll(async () => {
            await dummyServer.close();
            await serverProcessManager.killAllProcesses();
          }, 30e3);

          it('Netlify should still load the Stencil app properly', async () => {
            await page.goto(searchPageEndpoint, {
              waitUntil: 'networkidle2',
            });

            expect(await page.$(searchInterfaceSelector)).not.toBeNull();
          }, 60e3);
        });
      }
    }
  );
});
