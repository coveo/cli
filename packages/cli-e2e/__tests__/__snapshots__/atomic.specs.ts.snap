// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui:create:atomic when using an existing pageId (--pageId flag specified) the project has been generated properly 1`] = `
HashedFolder {
  "children": [
    HashedFolder {
      "children": [
        HashedFolder {
          "children": [
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "svywi4m+5JDQ38UPkngnE1Mjtdo=",
                  "name": "areEnvironmentVariablesSet.ts",
                },
                HashedFile {
                  "hash": "qJrXPWX6cL4N+TMvhr6+n05vNwk=",
                  "name": "generateToken.ts",
                },
                HashedFile {
                  "hash": "AG8M1wAglfYRRL/+Y987D25+YcA=",
                  "name": "token.spec.ts",
                },
                HashedFile {
                  "hash": "LxpAYAmM+yDe4mLtd3Rj6R9S30k=",
                  "name": "token.ts",
                },
              ],
              "hash": "qIYVWdq8oh9lCcu0H1X5XfQchcU=",
              "name": "token",
            },
          ],
          "hash": "sBgXQK+6u5b7XzuUi8zQquqUtaU=",
          "name": "functions",
        },
        HashedFile {
          "hash": "qBh2kK6C/tcpIA4kKT8j7bm10Kw=",
          "name": "jest.config.js",
        },
        HashedFile {
          "hash": "Q5B4b7qfWLJaHE/TZpt/2vLRgUk=",
          "name": "LICENSE",
        },
        HashedFile {
          "hash": "3RItgq2aQ177ZrvBxVNrLLaaRLc=",
          "name": "netlify.toml",
        },
        HashedFile {
          "hash": "JcgnYZDTiQ6EqMwKIuEh9x+kLlA=",
          "name": "README.md",
        },
        HashedFile {
          "hash": "RmeGeAs5hovLIHxguToYky0yJDY=",
          "name": "tsconfig.json",
        },
      ],
      "hash": "MVuHWG3A8rTxFYXfcqRQA/UkK7s=",
      "name": "lambda",
    },
    HashedFile {
      "hash": "RH58BiYWbvwHsjuAvGykfQPEhA0=",
      "name": "netlify.toml",
    },
    HashedFile {
      "hash": "sHQjSgNkkzGDXBmN+aw/aMx7O1g=",
      "name": "README.md",
    },
    HashedFolder {
      "children": [],
      "hash": "2jmj7l5rSw0yVb/vlWAYkK/YBwk=",
      "name": "scripts",
    },
    HashedFolder {
      "children": [
        HashedFolder {
          "children": [
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "YY+tyg0Mw+JNjYdifQfLqfDUASQ=",
                  "name": "results-manager.tsx",
                },
                HashedFile {
                  "hash": "XRPU8q6nh23adL66sq+HEqg+lCY=",
                  "name": "template-1.html",
                },
              ],
              "hash": "jh83V2cKEQIWFje8KVLl+s96OXA=",
              "name": "results-manager",
            },
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "XemX7AoOihA4SgQsvWtcStjeiGo=",
                  "name": "sample-component.css",
                },
                HashedFile {
                  "hash": "Hc10iNLZdetWJmslIN073kcgqIo=",
                  "name": "sample-component.tsx",
                },
              ],
              "hash": "a6roJ6GZ6LokDTmBQ+gUBFvibHw=",
              "name": "sample-component",
            },
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "0kZk+nhPZP6DNXEt1ds/y13pPn4=",
                  "name": "sample-result-component.css",
                },
                HashedFile {
                  "hash": "t5lGbTWEHgY4Bp2hhfFGNxq0bXU=",
                  "name": "sample-result-component.tsx",
                },
              ],
              "hash": "KiAmaY02cMdCGTIWhsuqogs7zKk=",
              "name": "sample-result-component",
            },
          ],
          "hash": "fzMFT63mCvo5uu2exrKKhI2gGM8=",
          "name": "components",
        },
        HashedFile {
          "hash": "B8YfCYuao2stXw64J6gW1TONNVg=",
          "name": "components.d.ts",
        },
        HashedFile {
          "hash": "/MC1IvlmPW0zAgfo82n8CnBDMrA=",
          "name": "html.d.ts",
        },
        HashedFile {
          "hash": "hlBUDBUrW6Vub74g68bUqfs3OIM=",
          "name": "index.ts",
        },
        HashedFolder {
          "children": [],
          "hash": "2jmj7l5rSw0yVb/vlWAYkK/YBwk=",
          "name": "pages",
        },
        HashedFolder {
          "children": [
            HashedFile {
              "hash": "61GVF7l8bCVXW00H2rtInTKwoqY=",
              "name": "index.css",
            },
          ],
          "hash": "FQZ6fv/Z/8DhwZWuUbrLwgTt6r8=",
          "name": "style",
        },
      ],
      "hash": "MyBnsgmKg3qrQRu4sQZ7FfT7jgQ=",
      "name": "src",
    },
    HashedFile {
      "hash": "jVuOZ6BrKT9HUer88GsGfeySq/w=",
      "name": "start-netlify.mjs",
    },
    HashedFile {
      "hash": "XiXrxQ8C2vt88gPhq3hGtPIJ5Yg=",
      "name": "tsconfig.json",
    },
  ],
  "hash": "3hfx19GW+7poeUXP9QQZn5h9nn0=",
  "name": "normalizedDir",
}
`;

exports[`ui:create:atomic when using an existing pageId (pageId not specified) the project has been generated properly 1`] = `
HashedFolder {
  "children": [
    HashedFolder {
      "children": [
        HashedFolder {
          "children": [
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "svywi4m+5JDQ38UPkngnE1Mjtdo=",
                  "name": "areEnvironmentVariablesSet.ts",
                },
                HashedFile {
                  "hash": "qJrXPWX6cL4N+TMvhr6+n05vNwk=",
                  "name": "generateToken.ts",
                },
                HashedFile {
                  "hash": "AG8M1wAglfYRRL/+Y987D25+YcA=",
                  "name": "token.spec.ts",
                },
                HashedFile {
                  "hash": "LxpAYAmM+yDe4mLtd3Rj6R9S30k=",
                  "name": "token.ts",
                },
              ],
              "hash": "qIYVWdq8oh9lCcu0H1X5XfQchcU=",
              "name": "token",
            },
          ],
          "hash": "sBgXQK+6u5b7XzuUi8zQquqUtaU=",
          "name": "functions",
        },
        HashedFile {
          "hash": "qBh2kK6C/tcpIA4kKT8j7bm10Kw=",
          "name": "jest.config.js",
        },
        HashedFile {
          "hash": "Q5B4b7qfWLJaHE/TZpt/2vLRgUk=",
          "name": "LICENSE",
        },
        HashedFile {
          "hash": "3RItgq2aQ177ZrvBxVNrLLaaRLc=",
          "name": "netlify.toml",
        },
        HashedFile {
          "hash": "JcgnYZDTiQ6EqMwKIuEh9x+kLlA=",
          "name": "README.md",
        },
        HashedFile {
          "hash": "RmeGeAs5hovLIHxguToYky0yJDY=",
          "name": "tsconfig.json",
        },
      ],
      "hash": "MVuHWG3A8rTxFYXfcqRQA/UkK7s=",
      "name": "lambda",
    },
    HashedFile {
      "hash": "RH58BiYWbvwHsjuAvGykfQPEhA0=",
      "name": "netlify.toml",
    },
    HashedFile {
      "hash": "sHQjSgNkkzGDXBmN+aw/aMx7O1g=",
      "name": "README.md",
    },
    HashedFolder {
      "children": [],
      "hash": "2jmj7l5rSw0yVb/vlWAYkK/YBwk=",
      "name": "scripts",
    },
    HashedFolder {
      "children": [
        HashedFolder {
          "children": [
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "YY+tyg0Mw+JNjYdifQfLqfDUASQ=",
                  "name": "results-manager.tsx",
                },
                HashedFile {
                  "hash": "XRPU8q6nh23adL66sq+HEqg+lCY=",
                  "name": "template-1.html",
                },
              ],
              "hash": "jh83V2cKEQIWFje8KVLl+s96OXA=",
              "name": "results-manager",
            },
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "XemX7AoOihA4SgQsvWtcStjeiGo=",
                  "name": "sample-component.css",
                },
                HashedFile {
                  "hash": "Hc10iNLZdetWJmslIN073kcgqIo=",
                  "name": "sample-component.tsx",
                },
              ],
              "hash": "a6roJ6GZ6LokDTmBQ+gUBFvibHw=",
              "name": "sample-component",
            },
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "0kZk+nhPZP6DNXEt1ds/y13pPn4=",
                  "name": "sample-result-component.css",
                },
                HashedFile {
                  "hash": "t5lGbTWEHgY4Bp2hhfFGNxq0bXU=",
                  "name": "sample-result-component.tsx",
                },
              ],
              "hash": "KiAmaY02cMdCGTIWhsuqogs7zKk=",
              "name": "sample-result-component",
            },
          ],
          "hash": "fzMFT63mCvo5uu2exrKKhI2gGM8=",
          "name": "components",
        },
        HashedFile {
          "hash": "2/Ssowtzkh/hx4e9vrMi11FFp1Q=",
          "name": "components.d.ts",
        },
        HashedFile {
          "hash": "/MC1IvlmPW0zAgfo82n8CnBDMrA=",
          "name": "html.d.ts",
        },
        HashedFile {
          "hash": "hlBUDBUrW6Vub74g68bUqfs3OIM=",
          "name": "index.ts",
        },
        HashedFolder {
          "children": [],
          "hash": "2jmj7l5rSw0yVb/vlWAYkK/YBwk=",
          "name": "pages",
        },
        HashedFolder {
          "children": [
            HashedFile {
              "hash": "61GVF7l8bCVXW00H2rtInTKwoqY=",
              "name": "index.css",
            },
          ],
          "hash": "FQZ6fv/Z/8DhwZWuUbrLwgTt6r8=",
          "name": "style",
        },
      ],
      "hash": "fTp/vtQ1keqpa0vWOA+VJlSEaFk=",
      "name": "src",
    },
    HashedFile {
      "hash": "jVuOZ6BrKT9HUer88GsGfeySq/w=",
      "name": "start-netlify.mjs",
    },
    HashedFile {
      "hash": "XiXrxQ8C2vt88gPhq3hGtPIJ5Yg=",
      "name": "tsconfig.json",
    },
  ],
  "hash": "qiQpjO78qNq7c1hobjxhbaa2vSY=",
  "name": "normalizedDir",
}
`;

exports[`ui:create:atomic when using an existing pageId (using the list prompt of available pages) the project has been generated properly 1`] = `
HashedFolder {
  "children": [
    HashedFile {
      "hash": "RH58BiYWbvwHsjuAvGykfQPEhA0=",
      "name": "netlify.toml",
    },
    HashedFile {
      "hash": "sHQjSgNkkzGDXBmN+aw/aMx7O1g=",
      "name": "README.md",
    },
    HashedFolder {
      "children": [
        HashedFile {
          "hash": "Jy9yA5W89BgTcZN4pVbSygA3w9U=",
          "name": "clean-up.js",
        },
        HashedFile {
          "hash": "8HGPil/NJ96KWhjo/uk9JBaq9us=",
          "name": "setup-lambda.mjs",
        },
        HashedFile {
          "hash": "SPggJ4gGqtRc5+8WHpeKd/gQcf0=",
          "name": "utils.js",
        },
      ],
      "hash": "0DHKkhnIXmyyA6198J3HAA9T0C0=",
      "name": "scripts",
    },
    HashedFolder {
      "children": [
        HashedFolder {
          "children": [
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "ErdFGgbxq+C0LOYeQvVh59O3JYA=",
                  "name": "results-manager.tsx",
                },
                HashedFile {
                  "hash": "/m87zRv/MiIstEldgz3NO4atbZM=",
                  "name": "template-1.html",
                },
              ],
              "hash": "L2wiyeP//J6E0Gbs5Yu1GFfxF1U=",
              "name": "results-manager",
            },
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "XemX7AoOihA4SgQsvWtcStjeiGo=",
                  "name": "sample-component.css",
                },
                HashedFile {
                  "hash": "YTj99Fegnant/k2ql9TPe2YMIKI=",
                  "name": "sample-component.tsx",
                },
              ],
              "hash": "q1Hnmgypio/LCK0csDi0uwZUZBg=",
              "name": "sample-component",
            },
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "0kZk+nhPZP6DNXEt1ds/y13pPn4=",
                  "name": "sample-result-component.css",
                },
                HashedFile {
                  "hash": "WTNvX4aKbAWuOwDGCn8De81zB24=",
                  "name": "sample-result-component.tsx",
                },
              ],
              "hash": "CUouJ7KRaYkS35p9RQ6d7HwrEwQ=",
              "name": "sample-result-component",
            },
          ],
          "hash": "IzpnK5rbJT7d4vrimnemc+St5pI=",
          "name": "components",
        },
        HashedFile {
          "hash": "zOAUK2vUx2PHWaGj0Rg/i3tFp7Q=",
          "name": "components.d.ts",
        },
        HashedFile {
          "hash": "YVuwXfIZfjfeVyLQ6cMy3kUqC3A=",
          "name": "html.d.ts",
        },
        HashedFile {
          "hash": "VYQYwWUyFPlPYQDod3YJsXVOEiQ=",
          "name": "index.ts",
        },
        HashedFolder {
          "children": [],
          "hash": "2jmj7l5rSw0yVb/vlWAYkK/YBwk=",
          "name": "pages",
        },
        HashedFolder {
          "children": [
            HashedFile {
              "hash": "ygdkeGA2hXdvtcqJyetblv/fdrw=",
              "name": "index.css",
            },
          ],
          "hash": "lRdUDVao1PTEjkh/JyuWMC0NvKk=",
          "name": "style",
        },
      ],
      "hash": "BsaZYMD1wDqcmu1eKmK0uFjxpKk=",
      "name": "src",
    },
    HashedFile {
      "hash": "XRdANxlMd+KVd0HuNLxu+WqpRu0=",
      "name": "start-netlify.mjs",
    },
    HashedFile {
      "hash": "XiXrxQ8C2vt88gPhq3hGtPIJ5Yg=",
      "name": "tsconfig.json",
    },
  ],
  "hash": "AOiP/eQTOWv66H9p5nIJj6FgxjA=",
  "name": "normalizedDir",
}
`;

exports[`ui:create:atomic when using the default page config (pageId not specified) the project has been generated properly 1`] = `
HashedFolder {
  "children": [
    HashedFile {
      "hash": "RH58BiYWbvwHsjuAvGykfQPEhA0=",
      "name": "netlify.toml",
    },
    HashedFile {
      "hash": "sHQjSgNkkzGDXBmN+aw/aMx7O1g=",
      "name": "README.md",
    },
    HashedFolder {
      "children": [
        HashedFile {
          "hash": "Jy9yA5W89BgTcZN4pVbSygA3w9U=",
          "name": "clean-up.js",
        },
        HashedFile {
          "hash": "8HGPil/NJ96KWhjo/uk9JBaq9us=",
          "name": "setup-lambda.mjs",
        },
        HashedFile {
          "hash": "SPggJ4gGqtRc5+8WHpeKd/gQcf0=",
          "name": "utils.js",
        },
      ],
      "hash": "0DHKkhnIXmyyA6198J3HAA9T0C0=",
      "name": "scripts",
    },
    HashedFolder {
      "children": [
        HashedFolder {
          "children": [
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "X8T+MmiPJFNu/lwXh6d29i/YhYo=",
                  "name": "results-manager.tsx",
                },
                HashedFile {
                  "hash": "OMU1pEfGxxywcokm8aFD8znPyaI=",
                  "name": "template-1.html",
                },
              ],
              "hash": "W1lEfKcMoS9W1dKtbS/qCjUkWkQ=",
              "name": "results-manager",
            },
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "XemX7AoOihA4SgQsvWtcStjeiGo=",
                  "name": "sample-component.css",
                },
                HashedFile {
                  "hash": "YTj99Fegnant/k2ql9TPe2YMIKI=",
                  "name": "sample-component.tsx",
                },
              ],
              "hash": "q1Hnmgypio/LCK0csDi0uwZUZBg=",
              "name": "sample-component",
            },
            HashedFolder {
              "children": [
                HashedFile {
                  "hash": "0kZk+nhPZP6DNXEt1ds/y13pPn4=",
                  "name": "sample-result-component.css",
                },
                HashedFile {
                  "hash": "WTNvX4aKbAWuOwDGCn8De81zB24=",
                  "name": "sample-result-component.tsx",
                },
              ],
              "hash": "CUouJ7KRaYkS35p9RQ6d7HwrEwQ=",
              "name": "sample-result-component",
            },
          ],
          "hash": "bCmtp7oSx57hExkjNmsDKz/kH/I=",
          "name": "components",
        },
        HashedFile {
          "hash": "zOAUK2vUx2PHWaGj0Rg/i3tFp7Q=",
          "name": "components.d.ts",
        },
        HashedFile {
          "hash": "YVuwXfIZfjfeVyLQ6cMy3kUqC3A=",
          "name": "html.d.ts",
        },
        HashedFile {
          "hash": "VYQYwWUyFPlPYQDod3YJsXVOEiQ=",
          "name": "index.ts",
        },
        HashedFolder {
          "children": [],
          "hash": "2jmj7l5rSw0yVb/vlWAYkK/YBwk=",
          "name": "pages",
        },
        HashedFolder {
          "children": [
            HashedFile {
              "hash": "Dp2EpyDmwkxrxWxZxv0cM41ulEs=",
              "name": "index.css",
            },
          ],
          "hash": "aAm3g7fNVc/yWUitXMGFxqC9GIg=",
          "name": "style",
        },
      ],
      "hash": "Aoo+O/XADnvB2BB2L7rxMNSElVQ=",
      "name": "src",
    },
    HashedFile {
      "hash": "XRdANxlMd+KVd0HuNLxu+WqpRu0=",
      "name": "start-netlify.mjs",
    },
    HashedFile {
      "hash": "XiXrxQ8C2vt88gPhq3hGtPIJ5Yg=",
      "name": "tsconfig.json",
    },
  ],
  "hash": "LollTEIcqbvmu3s05590vR40Erc=",
  "name": "normalizedDir",
}
`;
